ex12.c

```diff
@@ -0,0 +1,297 @@
+	PC_int(PC_get(conf, ".parallelism.width" ), &longval); psize[1] = longval;
+	
+	// check the configuration is coherent
+	assert(global_size[0]%psize[0]==0);
+	assert(global_size[1]%psize[1]==0);
+	assert(psize[1]*psize[0] == psize_1d);
+	
+	// compute the local data-size with space for ghosts and boundary constants
+	dsize[0] = global_size[0]/psize[0] + 2;
+	dsize[1] = global_size[1]/psize[1] + 2;
+	
+	// create a 2D Cartesian MPI communicator & get our coordinate (rank) in it
+	int cart_period[2] = { 1, 1 };
+	MPI_Comm cart_comm; MPI_Cart_create(main_comm, 2, psize, cart_period, 1, &cart_comm);
+	MPI_Cart_coords(cart_comm, pcoord_1d, 2, pcoord);
+	
+	// allocate memory for the double buffered data
+	double(*cur)[dsize[1]]  = malloc(sizeof(double)*dsize[1]*dsize[0]);
+	double(*next)[dsize[1]] = malloc(sizeof(double)*dsize[1]*dsize[0]);
+	
+	// initialize the value of switch for each iterations
+	int switch_iter_values[10] = {20, 35, 50, 55, 60, 35, 25, 20, 15, 60 };
+
+	// open the file should_output.dat
+	open_file();
+
+	// initialize the data content
+	PDI_event("initialization");
+	init(cur);
+	
+	// our loop counter so as to be able to use it outside the loop
+	int ii=0;
+	// share useful configuration bits with PDI
+	PDI_expose("pcoord",     pcoord, PDI_OUT);
+	PDI_expose("dsize",      dsize,  PDI_OUT);
+	PDI_expose("psize",      psize,  PDI_OUT);
+	
+	// value of switch inside the iteration loop
+	int tmp_switch;
+	// pointer to the value of should_output
+	bool *should_output;
+
+	// get the initial value of should_output
+	PDI_access("should_output",  (void **)&should_output,  PDI_IN);
+	PDI_release("should_output");
+	if(pcoord_1d == 0) printf("initial value: should_output = %d \n", *should_output);
+	
+	// the main loop
+	for (; ii<10; ++ii) {
+		// set value of switch inside the iteration loop
+		tmp_switch = switch_iter_value[ii];
+		if(pcoord_1d == 0) printf("iter = %d, switch = %d\n", ii, tmp_switch);
+		PDI_expose("switch", &tmp_switch, PDI_OUT);
+
+		// get should_output value and write to the file should_output.dat
+		PDI_access("should_output",  (void **)&should_output,  PDI_IN);
+		PDI_release("should_output");
+		if(pcoord_1d == 0) printf("iter = %d, should_output = %d \n", ii, *should_output);
+		if(pcoord_1d == 0) fprintf(pFile2, "%d\t%d\t%d\n", ii, tmp_switch, *should_output);
+
+		// share the loop counter & main field at each iteration
+		PDI_multi_expose("loop",
+				"ii",         &ii, PDI_OUT,
+				"main_field", cur, PDI_OUT,
+				NULL);
+		
+		// compute the values for the next iteration
+		iter(cur, next);
+		
+		// exchange data with the neighbours
+		exchange(cart_comm, next);
+		
+		// swap the current and next values
+		double (*tmp)[dsize[1]] = cur; cur = next; next = tmp;
+	}
+	// finally share the main field as well as the loop counter after the loop
+	PDI_multi_expose("finalization",
+	        "ii",         &ii, PDI_OUT,
+	        "main_field", cur, PDI_OUT,
+	        NULL);
```

Comment:
```suggestion
		PDI_multi_expose("loop",
				 "ii",         &ii, PDI_OUT,
				 "main_field", cur, PDI_OUT,
				 NULL);
		
		// compute the values for the next iteration
		iter(cur, next);
		
		// exchange data with the neighbours
		exchange(cart_comm, next);
		
		// swap the current and next values
		double (*tmp)[dsize[1]] = cur; cur = next; next = tmp;
	}
	// finally share the main field as well as the loop counter after the loop
	PDI_multi_expose("finalization",
	        	 "ii",         &ii, PDI_OUT,
	        	 "main_field", cur, PDI_OUT,
	        	 NULL);
```
Unify multi_expose indent.